third_party/amd/lib/TritonAMDGPUTransforms/ConvertToBufferOps.cpp

```diff
@@ -239,7 +239,37 @@ bool canUseBufferOps(Value ptr,
     return false;
   LDBG("32 bit offset");
 
-  return verifyNonNegativeExpr(offset, assumptions, std::move(solver));
+  bool nonNegativeOffset = verifyNonNegativeExpr(offset, assumptions, solver);
+  // It's not sufficient to have a non-negative offset. We need to ensure this
+  // offset is non-negative in BYTES.
+  Type elementType = getElementTypeOrSelf(addPtrOp);
+  while (elementType.isa<triton::PointerType>()) {
+    elementType = dyn_cast<triton::PointerType>(elementType).getPointeeType();
+  }
+  const int valueElemNBits = std::max(8u, elementType.getIntOrFloatBitWidth());
+  const int elementByteWidth = valueElemNBits / 8;
+  if (nonNegativeOffset) {
+    // Check for initialization because not all code uses the RangeAnalysis
+    // at this time.
+    // TODO: Remove!!!!
```

Comment:
Most of this code is to guarantee we say "use BufferOps because `dataflow::staticallyNonNegative(*solver, offset)` returns True. If we simplify `verifyNonNegativeExpr` to only use RangeAnalysis then none of this is required any longer.